pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,198 @@
+// Designed to run async.
+// TODO:
+// 	- Add channel param for quit (better error handling as well)
+// 	- Because there are 3 possible configs, there are 3 possible healthcheck paths, and all of them
+// 		are defined here. Therefore, there is quite a bit of duplicate code in success/fail threshold
+// 		and healthChannel sending. This code should be better.
+// 	- Should the CheckFrequency Delay be a time.Ticker?
+func (w *WorkloadHealthChecker) PerformApplicationHealthCheck(notifyHealthChange chan *discovery.DiscoveryRequest) {
+	// delay before starting probes.
+	time.Sleep(w.config.InitialDelay)
+
+	numSuccess, numFail := 0, 0
+	if w.config.CheckType == HTTPHealthCheck {
+		for {
+			if code, err := httpCheck(w.config.HTTPConfig, w.config.ProbeTimeout); code >= 200 && code <= 299 {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(code),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+			// should this be a time.Ticker?
+			time.Sleep(w.config.CheckFrequency)
+		}
+	}
+
+	if w.config.CheckType == TCPHealthCheck {
+		for {
+			if err := tcpCheck(w.config.TCPConfig, w.config.ProbeTimeout); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(500),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+		}
+	}
+
+	if w.config.CheckType == ExecHealthCheck {
+		for {
+			if err := execCheck(w.config.ExecConfig); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				} else {
+					numFail++
+					if numFail == w.config.FailThresh {
+						notifyHealthChange <- &discovery.DiscoveryRequest{
+							TypeUrl: HealthInfoTypeURL,
+							ErrorDetail: &status.Status{
+								Code:    int32(500),
+								Message: err.Error(),
+							},
+						}
+						numFail = 0
+					}
+				}
+			}
+		}
+	}
+}
+
+// httpCheck performs a http get to a given endpoint with a timeout, and returns
+// the status and error.
+func httpCheck(config HTTPHealthCheckConfig, timeout time.Duration) (int, error) {
+	client := http.Client{
+		Timeout: timeout,
+	}
+	url := fmt.Sprintf("%s://localhost:%v/%s", config.Scheme, config.Port, config.Path)
```

Comment:
How does scheme=https work? I think we need to add some tls configuration, in particular insecureSkipVerify?

Reply:
yep. added.